docs/fenrir.1
.\"t
.\" Automatically generated by Pandoc 1.19.2.1
.\"
.TH "" "" "" "" ""
.hy
.SH Fenrir User Manual
.PP
Fenrir is a modern command line screen reader written in Python3.
.PP
It has a modular structure, a flexible based driver model, is highly
configurable and easy to customize and extend (see Developer
Manual (fenrir_development_manual)).
.PP
Please see the following pages for the current (fenrir_current_version)
and Git (fenrir_git_version) version of Fenrir.
.SH Support and Requirements
.PP
Fenrir requires several drivers to interact with the operating system.
.SS Speech Drivers
.PP
A speech driver is for communication with the text to speech system like
Speech\-Dispatcher (#SpeechDispatcher) or
Espeak (http://espeak.sourceforge.net).
\\ See section Speech (#Speech) in \[aq]\[aq]settings.conf\[aq]\[aq] for
more information.
.SS SpeechDispatcher
.PP
This driver is used by default.
It uses Speech\-dispatcher as its backend.
.PP
Dependencies:
.IP \[bu] 2
Speech\-dispatcher (installed and configured,
Documentation (https///devel.freebsoft.org/speechd#sec2))
.IP \[bu] 2
Python\-speechd
.SS Espeak
.PP
Uses Espeak via Python bindings.
.PP
Dependencies:
.IP \[bu] 2
Espeak or Espeak\-ng
.IP \[bu] 2
python\-espeak (https///launchpad.net/python-espeak)
.SS Generic
.PP
This invokes speech via a sub\-process.
This is almost the same as using the commandline.
The performance depends on the overhead of the speech synthesis
application but it is really flexible.
.PP
Dependencies:
.IP \[bu] 2
Espeak or Espeak\-ng
.PP
The Requirements are flexible, they depend on the configuration in
settings.conf.
.SS Dummy
.PP
this is just for debugging, logs are output to the screen and logged as
well.
.SS Sound Drivers
.PP
To play sound icons and similar.\\ See section Sound (#Sound) in
\[aq]\[aq]settings.conf\[aq]\[aq] for more information.
.SS Generic
.PP
This driver is used by default.
.PP
Dependencies:
.IP \[bu] 2
Sox (http://sox.sourceforge.net/) with opus support The Requirements are
flexible, they depend on the configuration in settings.conf.
.SS Gstreamer
.PP
if you prefer to use Gstreamer for sound output.
.PP
Dependencies:
.IP \[bu] 2
Gstreamer >= 1.x
.IP \[bu] 2
Glibc
.SS Dummy
.PP
this is just for debugging, logs are output to the screen and logged as
well.
.SS Input Drivers
.PP
Input drivers are to capture keyboard shortcuts issued to the screen
reader
.PD 0
.P
.PD
See section Keyboard (#Keyboard) in \[aq]\[aq]settings.conf\[aq]\[aq]
for more information.
.SS Evdev
.PP
This driver is used by default.
.PP
Evdev is the low level input device framework for Linux.
.PP
Dependencies:
.IP \[bu] 2
python\-evdev >=0.6.3
.IP \[bu] 2
pyudev
.IP \[bu] 2
loaded uinput kernel module
.IP \[bu] 2
exclusive access to the input devices Read permission to the following
files and services:
.IP \[bu] 2
/dev/input
.IP \[bu] 2
/dev/uinput
.SS Screen Drivers
.PP
The job of a screen driver is to get the information of current screen
content.\\ See section Screen (#Screen) in
\[aq]\[aq]settings.conf\[aq]\[aq] for more information.
.SS VCSA
.PP
This driver is used by default.
For Linux VCSA devices.
These exist on any current standard installation of Linux.
.PP
Dependencie s:
.IP \[bu] 2
python\-dbus Read permission to the following files and services (or run
as root):
.IP \[bu] 2
/sys/devices/virtual/tty/tty0/active
.IP \[bu] 2
/dev/tty[1 \- 64]
.IP \[bu] 2
/dev/vcsa[1 \- 64] (VCSA manpage (https///linux.die.net/man/4/vcsa))
.IP \[bu] 2
read Logind DBUS
.SS Braille Drivers
.PP
This is for Braille support.
Braille is currently a work in progress and is planned for the Fenrir
2.0 release.\\ See section Braille (#Braille) in
\[aq]\[aq]settings.conf\[aq]\[aq] for more information.
.SS BRLTTY
.PP
This driver is used by default.
It uses BrlTTY (brltty) to communicate with with a Braille device.
.PP
Dependencies:
.IP \[bu] 2
BrlTTY (configured and running,
Documentation (http://mielke.cc/brltty/doc/Manual-BRLTTY/English/BRLTTY.html))
.IP \[bu] 2
python\-brlapi (configured,
Documentation (http://mielke.cc/brltty/doc/Manual-BrlAPI/English/BrlAPI.html))
## Currently supported platforms
.PP
Currently Fenrir completely supports the following Platforms:
.IP \[bu] 2
Linux TTY Support for further Systems are planned.
.SH Installation
.PP
Fenrir can run without installation.
It just requires the dependencies are installed first.
.PP
We recommend to try it out before installation to be sure everything
works and prevent yourself from experiencing a non\-talking environment.
.SS Try Out
.PP
Fenrir does not require installation.
You can try it and make sure everything works before you decide to
install.
In this way you can be sure that your system doesnt break or stop
talking.
for that you can just grab the code and run as root
\[aq]\[aq]src/fenrir/fenrir\[aq]\[aq] (in foreground) or
\[aq]\[aq]src/fenrir/fenrir\-daemon\[aq]\[aq] (in background, used by
systemd for autostart)
.SS Install it
.SS Documented operating systems
.SS Arch Linux
.PP
For Arch there are PKGBUILDs in the AUR:
.IP \[bu] 2
fenrir (https///aur.archlinux.org/packages/fenrir/)
.IP \[bu] 2
fenrir\-git (https///aur.archlinux.org/packages/fenrir-git/)
.SS Manual
.IP "1." 3
Download the latest stable version from the
Fenrir\-Project (https///linux-a11y.org/index.php?page=fenrir-screenreader)
site.
.IP "2." 3
Unpack the archive
.IP "3." 3
Check the needed Dependencys by running
check\-dependencys.py (https///github.com/chrys87/fenrir/blob/master/check-dependencies.py)
script
.IP "4." 3
install the missing dependencies an standard installation requires the
following:
.RS 4
.IP \[bu] 2
python3 >= 3.3 (and all the following is needed for python3 )
.IP \[bu] 2
python3\-speechd (screen)
.IP \[bu] 2
python3\-dbus (screen)
.IP \[bu] 2
python3\-evdev >= 0.6.4(input)
.IP \[bu] 2
python3\-daemonize (background service)
.IP \[bu] 2
python3\-brlapi (braille)
.IP \[bu] 2
python3\-pyenchant (spellchecker)
.IP \[bu] 2
your language for aspell (aspell\-\f[C]<lang>\f[]) (spellchecker)
.IP \[bu] 2
sox (sound)
.IP \[bu] 2
For an individual installation see Support and
Requirements (#Support%20and%20Requirements) or consult the
Readme (https///github.com/chrys87/fenrir/blob/master/README.md))
.RE
.IP "5." 3
run "install.sh" as root
.PP
this installs Fenrir as the following
.IP
.nf
\f[C]
*\ Application:\[aq]\[aq]/opt/fenrir\[aq]\[aq]
*\ Settings:\[aq]\[aq]/etc/fenrir\[aq]\[aq]
*\ Sound\ Icons:\[aq]\[aq]/usr/share/fenrirscreenreader/\[aq]\[aq]
\f[]
.fi
.PP
to remove Fenrir just run uninstall.sh as root
.SS Git
.PP
if you want to get the latest code you can use git to get a development
snapshot:
.IP
.nf
\f[C]
git\ clone\ https://github.com/chrys87/fenrir.git
\f[]
.fi
.SS Auto Start
.PP
To start Fenrir once: systemctl start fenrir
.PP
To enable auto start on system boot: systemctl enable fenrir
.SH First Steps
.PP
If you are using Fenrir for the first time you may want to take a look
at these resources:
.IP \[bu] 2
Keybindings (#Keybindings)
.IP \[bu] 2
Tutorial Mode (#Tutorial%20Mode)
.SH Features
.SS Commands
.SS Keybindings
.PP
Normal commands can be invoked in two ways: 1.
Using a Metakey (FenrirKey (#Fenrir%20Key)) 2.
Shortcuts with a single key
.PP
See section Keyboard (#Keyboard) in \[aq]\[aq]settings.conf\[aq]\[aq]
for more information.
#### Fenrir Key
.PP
The Fenrir Key is for invoking screen reader commands.
Fenrir can utilize more than one FenrirKey at the same time.
By default the following keys are used: 1.
Insert 2.
KeyPad Insert 3.
Meta (Super, Windows)
.SS Script Key
.PP
To invoke "Scripts" the Script Key is mandatory.
The shortcut is encoded in the filename of the script.
See Scripting (#Scripting) #### Desktop Layout
.PP
.TS
tab(@);
l l.
T{
Shortcut
T}@T{
Command
T}
_
T{
FenrirKey + H
T}@T{
toggle tutorial mode (#Tutorial%20Mode)
T}
T{
CTRL
T}@T{
shut up (interrupts speech) (#shut%20up)
T}
T{
FenrirKey + KeyPad 9
T}@T{
reviews bottom (#review%20bottom)
T}
T{
FenrirKey + KeyPad 7
T}@T{
reviews top (#review%20top)
T}
T{
KeyPad 8
T}@T{
reviews current line (#review%20current%20line)
T}
T{
KeyPad 7
T}@T{
reviews previous line (#review%20previous%20line)
T}
T{
KeyPad 9
T}@T{
reviews next line (#review%20next%20line)
T}
T{
FenrirKey + KeyPad 4
T}@T{
reviews line beginning (#review%20line%20beginning)
T}
T{
FenrirKey + KeyPad 6
T}@T{
reviews line ending (#review%20line%20ending)
T}
T{
FenrirKey + KeyPad 1
T}@T{
reviews line first character (#review%20line%20first%20character)
T}
T{
FenrirKey + KeyPad 3
T}@T{
reviews line last character (#review%20line%20last%20character)
T}
T{
FenrirKey + Alt + 1
T}@T{
presents first line (#present%20first%20line)
T}
T{
FenrirKey + Alt + 2
T}@T{
presents last line (#present%20last%20line)
T}
T{
KeyPad 5
T}@T{
reviews current word (#review%20current%20word)
T}
T{
KeyPad 4
T}@T{
reviews previous word (#review%20previous%20word)
T}
T{
KeyPad 6
T}@T{
reviews next word (#review%20next%20word)
T}
T{
FenrirKey + Shift + KeyPad 5
T}@T{
reviews current word phonetic (#review%20current%20word%20phonetic)
T}
T{
FenrirKey + Shift + KeyPad 4
T}@T{
reviews previous word phonetic (#review%20previous%20word%20phonetic)
T}
T{
FenrirKey + Shift + KeyPad 6
T}@T{
reviews next word phonetic (#review%20next%20word%20phonetic)
T}
T{
KeyPad 2
T}@T{
reviews current char (#review%20current%20character)
T}
T{
KeyPad 1
T}@T{
reviews previous char (#review%20previous%20character)
T}
T{
KeyPad 3
T}@T{
reviews next char (#review%20next%20character)
T}
T{
FenrirKey + Shift + KeyPad 2
T}@T{
reviews current character
phonetic (#review%20current%20character%20phonetic)
T}
T{
FenrirKey + Shift + KeyPad 1
T}@T{
reviews previous character
phonetic (#review%20previous%20character%20phonetic)
T}
T{
FenrirKey + Shift + KeyPad 3
T}@T{
reviews next character phonetic (#review%20next%20character%20phonetic)
T}
T{
FenrirKey + CTRL + KeyPad 8
T}@T{
reviews up (#review%20up)
T}
T{
FenrirKey + CTRL + KeyPad 2
T}@T{
reviews down (#review%20down)
T}
T{
FenrirKey + KeyPad dot
T}@T{
exit review (#exit%20review)
T}
T{
KeyPad dot
T}@T{
cursor position (#cursor%20position)
T}
T{
FenrirKey + I
T}@T{
indent curr line (#indent%20current%20line)
T}
T{
FenrirKey + KeyPad 5
T}@T{
current screen (#current%20screen)
T}
T{
FenrirKey + KeyPad 8
T}@T{
current screen before cursor (#current%20screen%20before%20cursor)
T}
T{
FenrirKey + KeyPad 2
T}@T{
current screen after cursor (#current%20screen%20after%20cursor)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
cursor read to end of line (#cursor%20read%20to%20end%20of%20line)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
cursor column (#cursor%20column)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
cursor line number (#cursor%20line%20number)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
Braille flush (#braille%20flush)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
Braille pan left (#braille%20pan%20left)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
Braille pan right (#braille%20pan%20right)
T}
T{
\f[C]<Unbound>\f[]
T}@T{
Braille return to cursor (#braille%20return%20to%20cursor)
T}
T{
FenrirKey + CTRL + 1
T}@T{
clear bookmark 1 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 1
T}@T{
set bookmark 1 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 1
T}@T{
bookmark 1 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 2
T}@T{
clear bookmark 2 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 2
T}@T{
set bookmark 2 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 2
T}@T{
bookmark 2 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 3
T}@T{
clear bookmark 3 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 3
T}@T{
set bookmark 3 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 3
T}@T{
bookmark 3 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 4
T}@T{
clear bookmark 4 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 4
T}@T{
set bookmark 4 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 4
T}@T{
bookmark 4 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 5
T}@T{
clear bookmark 5 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 5
T}@T{
set bookmark 5 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 5
T}@T{
bookmark 5 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 6
T}@T{
clear bookmark 6 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 6
T}@T{
set bookmark 6 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 6
T}@T{
bookmark 6 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 7
T}@T{
clear bookmark 7 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 7
T}@T{
set bookmark 7 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 7
T}@T{
bookmark 7 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 8
T}@T{
clear bookmark 8 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 8
T}@T{
set bookmark 8 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 8
T}@T{
bookmark 8 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 9
T}@T{
clear bookmark 9 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 9
T}@T{
set bookmark 9 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 9
T}@T{
bookmark 9 (#read%20Bookmark%20X)
T}
T{
FenrirKey + CTRL + 0
T}@T{
clear bookmark 10 (#clear%20Bookmark%20X)
T}
T{
FenrirKey + Shift + 0
T}@T{
set bookmark 10 (#set%20Bookmark%20X)
T}
T{
FenrirKey + 0
T}@T{
bookmark 10 (#read%20Bookmark%20X)
T}
T{
FenrirKey + KeyPad Slash
T}@T{
set window application (#Create%20Window)
T}
T{
2 * FenrirKey + KeyPad Slash
T}@T{
clear window application (#Remove%20Window)
T}
T{
KeyPad Plus
T}@T{
read last incoming (#last%20incoming)
T}
T{
FenrirKey + F2
T}@T{
toggles braille (#toggle%20braille)
T}
T{
FenrirKey + F3
T}@T{
toggles sound (#toggle%20sound)
T}
T{
FenrirKey + F4
T}@T{
toggles speech (#toggle%20speech)
T}
T{
KeyPad Enter
T}@T{
temporarily disables speech (#disable%20speech%20temporarily)
T}
T{
FenrirKey + CTRL + P
T}@T{
toggles punctuation level (#toggle%20punctuation%20level)
T}
T{
FenrirKey + RightBrace
T}@T{
toggle auto spell check (#toggle%20auto%20spell%20check)
T}
T{
FenrirKey + Backslash
T}@T{
toggles output (#toggle%20output)
T}
T{
FenrirKey + CTRL + E
T}@T{
toggles emoticons (#toggle%20emoticons)
T}
T{
FenrirKey + KeyPad Enter
T}@T{
toggles auto read (#toggle%20auto%20read)
T}
T{
FenrirKey + CTRL + T
T}@T{
toggles auto time (#toggle%20auto%20time)
T}
T{
FenrirKey + KeyPad ASTERISK
T}@T{
toggles highlight tracking (#toggle%20highlight%20tracking)
T}
T{
FenrirKey + Q
T}@T{
quits fenrir (#quit%20Fenrir)
T}
T{
FenrirKey + T
T}@T{
Announce time (#Time)
T}
T{
2 * FenrirKey + T
T}@T{
Announce date (#Date)
T}
T{
FenrirKey + S
T}@T{
spell check (#spell%20check)
T}
T{
2 * FenrirKey + S
T}@T{
add word to spell check (#add%20word%20to%20spell%20check)
T}
T{
FenrirKey + Shift + S
T}@T{
removes word from spell check (#removes%20word%20from%20spell%20check)
T}
T{
FenrirKey + Backspace
T}@T{
forward keypress (#forward%20keypress)
T}
T{
FenrirKey + Up
T}@T{
increase speech volume (#increase%20speech%20volume)
T}
T{
FenrirKey + Down
T}@T{
decrease speech volume (#decrease%20speech%20volume)
T}
T{
FenrirKey + Right
T}@T{
increase speech rate (#increase%20speech%20rate)
T}
T{
FenrirKey + Left
T}@T{
decrease speech rate (#decrease%20speech%20rate)
T}
T{
FenrirKey + Alt + Right
T}@T{
increase speech pitch (#increase%20speech%20pitch)
T}
T{
FenrirKey + Alt + Left
T}@T{
decrease speech pitch (#decrease%20speech%20pitch)
T}
T{
FenrirKey + Alt + Up
T}@T{
increase sound volume (#increase%20sound%20volume)
T}
T{
FenrirKey + Alt + Down
T}@T{
decrease sound volume (#decrease%20sound%20volume)
T}
T{
FenrirKey + CTRL + Shift + C
T}@T{
clears clipboard (#clear%20clipboard)
T}
T{
FenrirKey + Home
T}@T{
first clipboard (#first%20clipboard)
T}
T{
FenrirKey + End
T}@T{
last clipboard (#last%20clipboard)
T}
T{
FenrirKey + PageUp
T}@T{
previous clipboard (#previous%20clipboard)
T}
T{
FenrirKey + PageDown
T}@T{
next clipboard (#next%20clipboard)
T}
T{
FenrirKey + Shift + C
T}@T{
current clipboard (#read%20current%20clipboard)
T}
T{
FenrirKey + C
T}@T{
copy marked text to clipboard (#copy%20marked%20to%20clipboard)
T}
T{
FenrirKey + V
T}@T{
paste clipboard contents (#paste%20clipboard)
T}
T{
FenrirKey + P
T}@T{
import clipboard from file (#import%20clipboard%20from%20file)
T}
T{
FenrirKey + Alt + Shift +C
T}@T{
export clipboard to file (#export%20clipboard%20to%20file)
T}
T{
FenrirKey + CTRL + Shift + X
T}@T{
remove marks (#Remove%20Marks)
T}
T{
FenrirKey + X
T}@T{
set mark (#Set%20mark)
T}
T{
FenrirKey + Shift + X
T}@T{
announce marked text (#Get%20text%20between%20marks)
T}
T{
Linux specific
T}@T{
T}
T{
\f[C]<Unbound>\f[]
T}@T{
export clipboard to X
T}
T{
FenrirKey + CTRL + Up
T}@T{
include Alsa volume
T}
T{
FenrirKey + CTRL + Down
T}@T{
decrease Alsa volume
T}
.TE
.SS Laptop Layout
.PP
.TS
tab(@);
l l.
T{
Shortcut
T}@T{
Command
T}
_
T{
FenrirKey + H
T}@T{
toggle tutorial mode (#Tutorial%20Mode)
T}
T{
CTRL
T}@T{
shut up (interrupts speech) (#shut%20up)
T}
T{
FenrirKey + Shift + O
T}@T{
reviews bottom (#review%20bottom)
T}
T{
FenrirKey + Shift + U
T}@T{
reviews top (#review%20top)
T}
T{
FenrirKey + I
T}@T{
reviews current line (#review%20current%20line)
T}
T{
FenrirKey + U
T}@T{
reviews previous line (#review%20previous%20line)
T}
T{
FenrirKey + O
T}@T{
reviews next line (#review%20next%20line)
T}
T{
FenrirKey + Shift + J
T}@T{
reviews line beginning (#review%20line%20beginning)
T}
T{
FenrirKey + Shift + L
T}@T{
reviews line ending (#review%20line%20ending)
T}
T{
FenrirKey + CTRL + J
T}@T{
reviews line first character (#review%20line%20first%20character)
T}
T{
FenrirKey + CTRL + L
T}@T{
reviews line last character (#review%20line%20last%20character)
T}
T{

Changelog.txt
# Version 1.5
- Doku: Write a user wiki 
  https://wiki.linux-a11y.org/doku.php?id=fenrir_user_manual&s[]=fenrir
  
- initial working setup.py

- leave review on typing 

- add dependency check (check-dependencys.py)

- Add nice dummy drivers as template or for debugging

- reimplement detection code for X11

- initial translate structure (manuelcortez Thanks!)  

- add a configurable place where you can place own commands or overwrite existing commands without need to change default code

- implement autodetection of plugged and unplugged input devices 

- implement speechdriver generic

- try to autodetect encoding (Easy for contribution) (Prototype "charmapTTY" in play zone)

Braille Support (WIP):
  - initial BrlTTY driver
  - detect device size via driver
  - output to braille device  
  - make flushMode configurable
  - make flushTimeout configurable
  - flush message after X seconds and show current line (review over text)
  - tweak current commands and output
  - command flush_braille
  - command for scroll left
  - command for scroll right  
  - create offset for scrolling
  - respect scrolling
  make cursor following configurable (brailleCursorTrackingMode)
    - cell
    - page  
  follow cursor while typing  
  brailleFocusMode:
    - review = priority to review

- move to an event based system

- add initial multithreading/ multiprocessing support

- support cli parameters
  - add cli parameter for debugging "-d"
  - add cli parameter to overwrite options "-o"
  - add cli parameter to specify an settings.conf "-s"

- list of bound commands in Tutorial Mode. speak name, binding and description.

# Version: 1.00

- move from VCS to VCSA and parese the Attributes
http://linux.die.net/man/4/vcsa
http://man.cx/vcsa(4)/de
http://manpages.org/display-vcsa/7
https://en.wikipedia.org/wiki/Virtual_console
every second byte is a attribute others are text. fast way: c[::2],c[1::2]
http://manpages.ubuntu.com/manpages/precise/de/man4/vcs.4.html
https://docs.python.org/3/library/fcntl.html
http://rodrigorivas.serveblog.net/en/imagenes-desde-vt-con-vcsa/
good doku:
http://angband.oook.cz/d/eyangband-052/src/main-vcs.c
http://manpages.ubuntu.com/manpages/trusty/man1/screader.1.html

- implement speechdriver espeak
https://github.com/relsi/python-espeak

- detect collumns in TTYs automaticaly. 
it seems we have this info in vcsa

- get current cursor
- shortcut handling
https://docs.python.org/2/library/termios.html
http://stackoverflow.com/questions/287757/pythons-configparser-unique-keys-per-section
0=down, 1=press, 2=hold
2KEY_SHIFT, 1KEY_A = say_current_line_cursor
- implement command structure
- implement speechdriver speechd
https://git.gnome.org/browse/orca/tree/src/orca/speech.py
https://git.gnome.org/browse/orca/tree/src/orca/speechdispatcherfactory.py
http://devel.freebsoft.org/doc/speechd/speech-dispatcher.html#Client-Programming

- autodetect current TTY maybe with (PAM or a sys folder)
cat /sys/devices/virtual/tty/tty0/active 
http://serverfault.com/questions/306854/how-to-find-out-the-currently-active-linux-virtual-terminal-while-connected-via

- Input 
http://python-evdev.readthedocs.io/en/latest/tutorial.html
http://stackoverflow.com/questions/12384772/how-can-i-capture-mouseevents-and-keyevents-using-python-in-background-on-linux
maybe TTY in RAW MODE

- Settings (make it configureable)
- improve differ speed
- lock mechanism for threads
- restructure loops to listen for events
  inputloop -> does block with an select
  commands -> a new thread should spawned from inputloop
  updatescreen -> maybe we could watch it with inotify vsca should support polling COMMENT: sadly not possible, poll events not fired as expected
  https://github.com/seb-m/pyinotify/wiki/Tutorial
<Example Code>
import pyinotify
import glob
class Identity(pyinotify.ProcessEvent):
    def process_default(self, event):
        p = event.pathname
        print(p)

wm = pyinotify.WatchManager()

notifier = pyinotify.Notifier(wm, default_proc_fun=Identity(), timeout=5)
wm.add_watch('/sys/devices/virtual/tty/tty0/active', pyinotify.IN_CLOSE_WRITE)
for file in list(glob.glob('/dev/vcsa[0-64]')):
    wm.add_watch(file, pyinotify.IN_CLOSE_WRITE)
    print(file)

try:
    while 1:
        notifier.process_events()
        if notifier.check_events( timeout=1000):
            notifier.read_events()
            print('events')
        else:
            print('timeout')
except KeyboardInterrupt:
    notifier.stop()
    print('fin')
</Example Code>

  https://www.infoq.com/articles/inotify-linux-file-system-event-monitoring
  https://github.com/seb-m/pyinotify/wiki/Tutorial
  http://www.saltycrane.com/blog/2010/04/monitoring-filesystem-python-and-pyinotify/
  
- add setting for ignore screens ( dont grab shortcuts from X or orca)    
- soundIcons
- performance tuning
- add sound volume
- convert volume to percent in config
- convert pitch to percent in config
- convert rate to percent in config
- make screenUpdate rate configurable
- default soundIcon theme (soundfiles)  
- debugging
- threading ReadContent, ReadShortcuts, executeCommands, listenNewTTYsForListen, controllThread (main)
- autoload plugins while starting
- implement sounddriver generic (use current sox and make it configurable)
- add setting for autodetect X
ps a -o tty,comm  | grep -e Xorg | grep -v "grep -e Xorg"
- respect window mode in differ (getwindow code is already in place)  
- parse punctuation setting file in conf/substitution

- implement commands
  curr_word
  curr_char
  next_word
  next_char
  prev_word
  prev_char
  enable_disable_speech #enable, disable speech
  enable_disable_braile #enable, disable braile
  enable_disable_sound #enable, disable sound 
  enable_disable_output #enable, disable speech, braile and sound
  next_clipboard
  prev_clipboard
  first_clipboard
  last_clipboard
  curr_clipboard
  paste_clipboard
  define_window
  remove_window  
  reset_review_on_screen_change
  remove_clipboard_marks
  copy_marked

src/fenrir/fenrir.py
#!/bin/env python3
# -*- coding: utf-8 -*-

# Fenrir TTY screen reader
# By Chrys, Storm Dragon, and contributers.

import os, sys, signal, time
import __main__

if not os.path.dirname(os.path.realpath(__main__.__file__)) in sys.path:
    sys.path.append(os.path.dirname(os.path.realpath(__main__.__file__)))
from core import i18n
from core import settingsManager
from core import debug

class fenrir():
    def __init__(self):
        try:
            self.environment = settingsManager.settingsManager().initFenrirConfig()
            if not self.environment:
                raise RuntimeError('Cannot Initialize. Maybe the configfile is not available or not parseable')
        except RuntimeError:
            raise
        self.environment['runtime']['outputManager'].presentText(_("Start Fenrir"), soundIcon='ScreenReaderOn', interrupt=True)          
        signal.signal(signal.SIGINT, self.captureSignal)
        signal.signal(signal.SIGTERM, self.captureSignal)
        self.wasCommand = False

    def proceed(self):
        while(self.environment['generalInformation']['running']):
            try:
                self.handleProcess()
            except Exception as e:
                self.environment['runtime']['debug'].writeDebugOut(str(e),debug.debugLevel.ERROR) 
        self.shutdown()

    def handleProcess(self):
        eventReceived = self.environment['runtime']['inputManager'].getInputEvent()
        startTime = time.time()    
        if not eventReceived:
            if not self.environment['runtime']['screenManager'].isSuspendingScreen():
                self.environment['runtime']['inputManager'].updateInputDevices()
        if eventReceived:
            self.prepareCommand()
            if not (self.wasCommand  or self.environment['generalInformation']['tutorialMode']) or  self.environment['runtime']['screenManager'].isSuspendingScreen():
                self.environment['runtime']['inputManager'].writeEventBuffer()
            if self.environment['runtime']['inputManager'].noKeyPressed():
                if self.wasCommand:
                        self.wasCommand = False   
                        self.environment['runtime']['inputManager'].clearEventBuffer()            
                if self.environment['generalInformation']['tutorialMode']:
                    self.environment['runtime']['inputManager'].clearEventBuffer()
                if self.environment['input']['keyForeward'] > 0:
                    self.environment['input']['keyForeward'] -=1
            self.environment['runtime']['screenManager'].update('onInput')                            
            self.environment['runtime']['commandManager'].executeDefaultTrigger('onInput')      
        else:
            self.environment['runtime']['screenManager'].update('onUpdate')
        if self.environment['runtime']['applicationManager'].isApplicationChange():
            self.environment['runtime']['commandManager'].executeDefaultTrigger('onApplicationChange')
            self.environment['runtime']['commandManager'].executeSwitchTrigger('onSwitchApplicationProfile', \
              self.environment['runtime']['applicationManager'].getPrevApplication(), \
              self.environment['runtime']['applicationManager'].getCurrentApplication())          
        
        if self.environment['runtime']['screenManager'].isScreenChange():    
            self.environment['runtime']['commandManager'].executeDefaultTrigger('onScreenChanged')             
        else:
            self.environment['runtime']['commandManager'].executeDefaultTrigger('onScreenUpdate')         
        #self.environment['runtime']['outputManager'].brailleText(flush=False)    
        self.handleCommands()
        #print(time.time()-startTime)       

    def prepareCommand(self):
        if self.environment['runtime']['screenManager'].isSuspendingScreen():
            self.wasCommand = False
            return        
        if self.environment['runtime']['inputManager'].noKeyPressed():
            return
        if self.environment['input']['keyForeward'] > 0:
            return
        shortcut = self.environment['runtime']['inputManager'].getCurrShortcut()        
        command = self.environment['runtime']['inputManager'].getCommandForShortcut(shortcut)        
        if len(self.environment['input']['prevDeepestInput']) <= len(self.environment['input']['currInput']):
            self.wasCommand = command != '' or self.environment['runtime']['inputManager'].isFenrirKeyPressed() or self.environment['runtime']['inputManager'].isScriptKeyPressed()    
        if command == '':
            return
            
        self.environment['runtime']['commandManager'].queueCommand(command)  

    def handleCommands(self): 
        if not self.environment['runtime']['commandManager'].isCommandQueued():
            return
        self.environment['runtime']['commandManager'].executeCommand( self.environment['commandInfo']['currCommand'], 'commands')

    def shutdownRequest(self):
        self.environment['generalInformation']['running'] = False

    def captureSignal(self, siginit, frame):
        self.shutdownRequest()

    def shutdown(self):
        self.environment['runtime']['outputManager'].presentText(_("Quit Fenrir"), soundIcon='ScreenReaderOff', interrupt=True)       
        for currManager in self.environment['generalInformation']['managerList']:
            if self.environment['runtime'][currManager]:
                self.environment['runtime'][currManager].shutdown()                      
                del self.environment['runtime'][currManager]
        self.environment = None

def main():
    app = fenrir()
    app.proceed()
    del app

if __name__ == "__main__":
    main()        
